granulate_utils/glogger/__init__.py

```diff
@@ -0,0 +1,189 @@
+    def dictify_record(self, record):
+        formatted_timestamp = datetime.utcfromtimestamp(record.created).isoformat()
+        return {
+            "severity": record.levelno,
+            "timestamp": formatted_timestamp,
+            "logger_name": record.name,
+            "message": record.message,
+        }
+
+    def _flush_loop(self) -> None:
+        self.last_flush_time = self.time_fn()
+        while not self.stop_event.is_set():
+            if self.should_flush():
+                self.flush()
+            self.stop_event.wait(1.0)
+
+    @property
+    def time_since_last_flush(self) -> float:
+        return self.time_fn() - self.last_flush_time
+
+    def should_flush(self) -> bool:
+        return (
+            self.messages_buffer.count > 0
+            and (self.messages_buffer.utilized >= self.flush_threshold)
+            or (self.time_since_last_flush >= self.flush_interval)
+        )
+
+    def flush(self) -> None:
+        # Allow configuring max_tries via class member. Alternatively we could decorate _send_logs in __init__ but that
+        # would be easier to miss and less flexible.
+        send = backoff.on_exception(backoff.expo, exception=RequestException, max_tries=self.max_send_tries)(
+            self._send_logs
+        )
+
+        self.last_flush_time = self.time_fn()
+        try:
+            batch, response = send()
+            response.raise_for_status()
+        except Exception:
+            logger.exception("Error posting to server")
+        else:
+            self.drop_sent_batch(batch)
+
+    def drop_sent_batch(self, sent_batch: Batch) -> None:
+        # we don't override createLock(), so lock is not None
+        with self.lock:  # type: ignore
+            dropped = self.messages_buffer.head_serial_no - sent_batch.head_serial_no
+            n = len(sent_batch.logs) - dropped
+            if n > 0:
+                self.messages_buffer.drop(n)
+
+    def _send_logs(self) -> Tuple[Batch, requests.Response]:
+        """
+        :return: (batch that was sent, response)
+        """
+        # Upon every retry we will remake the batch, in case we are able to batch more messages together.
+        batch = self.make_batch()
+        data = {
+            "batch_id": uuid.uuid4().hex,
+            "metadata": self.get_metadata(),
+            "logs": "<LOGS_JSON>",
+        }
+        # batch.logs is a list of json.dump()ed strings so ",".join() it into the final json string instead of json-ing
+        # the list.
+        body = json.dumps(data).replace('"<LOGS_JSON>"', f"[{','.join(batch.logs)}]")
+        response = self.session.post(
+            f"{self.scheme}://{self.server_address}/api/v1/logs",
+            data=body.encode("utf-8"),
+            headers={"Content-Type": "application/json"},
+            timeout=self.request_timeout,
+        )
+        return batch, response
+
+    def get_metadata(self) -> dict:
+        """Called to get metadata per batch."""
+        return {}
```

Comment:
Which metadata do you think should be here? The metadata is dynamic, so I expect each project to override `get_metadata` and add the fields it wants.

Reply:
Also, gprofiler places everything under `extra`, whereas other projects place fields directly in the metadata.

Reply:
Added example in readme.

Reply:
Well we're trying to make it common for all three - let's decide on a format for how to pass the metadata and do it. I do suggest using it like gProfiler does it as it works best right now IMO.

If I'd do it over, I think we should seperate between metadata and extra - metadata will be added to each log message under metadata field and should be constant for handler (so pass it in ctor). extra can be passed for each log invocation and should be added to the log under extra.

Since we also send a lot of common metadata (instance name, instance type, etc) maybe we want to place common default values which will be used for all of our agents (cpu type, host name, etc)